### PR TITLE
feat(js): introduce JavaScript package

### PR DESCRIPTION
**Pull request by shortcuts**
**Summary**

Translates React package to JavaScript, using Preact as an alias.

This low code solution will allow us to only maintain the React package and automatically have the changes reflected on the JavaScript one.

**Result**

A new `js-recommendations` package that works exactly like `react-recommendations`.

**Todo**
- [X] [Fix import](https://github.com/algolia/react-recommendations/blob/31208f322459ba26a726ff6bff76e6629e6d026a/examples/js-demo/app.js#L7), it is related to parcel itself.
- [ ] Update `README.md`

### CODE REVIEW

File: examples/js-demo/babel.config.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
```
Comment (francoischalifour) [inline examples/js-demo/babel.config.json]: We shouldn't need this file if we provide [`alias`](https://v2.parceljs.org/features/module-resolution/#aliases) in `package.json`.

File: package.json [hunk 2]
```diff
@@ -13,15 +13,21 @@
   "devDependencies": {
     "@algolia/client-search": "4.9.1",
     "@babel/cli": "7.13.0",
+    "@babel/plugin-transform-react-jsx": "^7.13.12",
```
Comment (francoischalifour) [inline package.json]: This dep should be pinned.

File: package.json [hunk 3]
```diff
@@ -13,15 +13,21 @@
...
+    "@rollup/plugin-json": "4.1.0",
+    "@rollup/plugin-node-resolve": "7.1.1",
+    "@rollup/plugin-replace": "2.3.3",
     "@testing-library/react": "11.2.6",
     "@types/jest": "^26.0.23",
     "@types/react": "^17.0.5",
     "@typescript-eslint/eslint-plugin": "2.19.0",
     "@typescript-eslint/parser": "2.19.0",
     "babel-eslint": "10.1.0",
+    "babel-plugin-module-resolver": "^4.1.0",
```
Comment (francoischalifour) [inline package.json]: This dep should be pinned.

File: package.json [hunk 4]
```diff
@@ -35,6 +41,12 @@
     "jest": "26.6.3",
     "lerna": "4.0.0",
     "prettier": "2.2.1",
-    "typescript": "4.2.4"
+    "typescript": "4.2.4",
+    "rollup": "1.31.0",
```
Comment (francoischalifour) [inline package.json]: Any reasons not to use Rollup v2?

File: packages/js-recommendations/rollup.config.js [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+if (!process.env.BUILD) {
+  throw new Error('The `BUILD` environment variable is required to build.');
+}
+
+const output = {
+  umd: {
+    file: 'dist/umd/index.js',
+    format: 'umd',
+    sourcemap: true,
+    name: '@algolia/js-recommendations',
```
Comment (francoischalifour) [inline packages/js-recommendations/rollup.config.js]: We can sync this value with `package.json`:

https://github.com/algolia/react-recommendations/blob/31208f322459ba26a726ff6bff76e6629e6d026a/packages/js-recommendations/package.json#L2

File: packages/js-recommendations/rollup.config.js [hunk 6]
```diff
@@ -0,0 +1,25 @@
+import { plugins } from '../../rollup.base.config';
+
+if (!process.env.BUILD) {
+  throw new Error('The `BUILD` environment variable is required to build.');
+}
+
+const output = {
+  umd: {
+    file: 'dist/umd/index.js',
```
Comment (francoischalifour) [inline packages/js-recommendations/rollup.config.js]: We can sync this value with `package.json`:

https://github.com/algolia/react-recommendations/blob/31208f322459ba26a726ff6bff76e6629e6d026a/packages/js-recommendations/package.json#L15

File: packages/js-recommendations/rollup.config.js [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    format: 'umd',
+    sourcemap: true,
+    name: '@algolia/js-recommendations',
+  },
+  esm: {
+    file: 'dist/esm/index.js',
```
Comment (francoischalifour) [inline packages/js-recommendations/rollup.config.js]: We can sync this value with `package.json`:

https://github.com/algolia/react-recommendations/blob/31208f322459ba26a726ff6bff76e6629e6d026a/packages/js-recommendations/package.json#L13

File: packages/js-recommendations/rollup.config.js [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+    format: 'es',
+    sourcemap: true,
+  },
+};
+
+export default {
+  input: 'src/index.ts',
```
Comment (francoischalifour) [inline packages/js-recommendations/rollup.config.js]: We can sync this value with `package.json`:

https://github.com/algolia/react-recommendations/blob/31208f322459ba26a726ff6bff76e6629e6d026a/packages/js-recommendations/package.json#L11

File: packages/js-recommendations/src/types/RenderProps.ts [hunk 9]
```diff
@@ -0,0 +1,4 @@
+export type RenderProps = {
```
Comment (francoischalifour) [inline packages/js-recommendations/src/types/RenderProps.ts]: The name `RenderProps` if a bit confusing. How about `DomProps`, `JsProps` or `EnvironmentProps`?

File: packages/react-recommendations/babel.config.js [hunk 10]
```diff
@@ -0,0 +1,25 @@
+module.exports = (api) => {
```
Comment (francoischalifour) [inline packages/react-recommendations/babel.config.js]: We can keep a single Babel config at the monorepo root and provide [`overrides`](https://babeljs.io/docs/en/options#overrides) for each package specificity.

**Comment by francoischalifour** [comment]
In a future PR we might want to add these APIs to each component:

```ts
type RelatedProductsApi<TObject extends ProductBaseRecord> = {
  update(props: RelatedProductsProps<TObject>): void;
  destroy(): void;
};
```

- `update` would call Preact's `render` with the new options
- `destroy` would call Preact's `render` with `null` to unmount the component

File: packages/js-recommendations/src/types/RenderProps.ts (hunk 9 above)
Comment (shortcuts) [reply packages/js-recommendations/src/types/RenderProps.ts]: `EnvironmentProps` seems appropriate to me!

Comment (shortcuts) [reply packages/js-recommendations/src/types/RenderProps.ts]: Done in https://github.com/algolia/react-recommendations/pull/6/commits/1690ca429f4dd755ae80fa7b2eca87aa6bc311c2

File: examples/js-demo/babel.config.json (hunk 1 above)
Comment (shortcuts) [reply examples/js-demo/babel.config.json]: Right! Done in https://github.com/algolia/react-recommendations/pull/6/commits/be611462b78ea8dc0603274a32af18d9b6964425

File: packages/react-recommendations/babel.config.js (hunk 10 above)
Comment (shortcuts) [reply packages/react-recommendations/babel.config.js]: Done in https://github.com/algolia/react-recommendations/pull/6/commits/956efe16d5ec7ded275cbac4540230c44dd0904c

File: packages/js-recommendations/rollup.config.js (hunk 8 above)
Comment (shortcuts) [reply packages/js-recommendations/rollup.config.js]: Done in https://github.com/algolia/react-recommendations/pull/6/commits/956efe16d5ec7ded275cbac4540230c44dd0904c

File: packages/js-recommendations/rollup.config.js (hunk 7 above)
Comment (shortcuts) [reply packages/js-recommendations/rollup.config.js]: Done in https://github.com/algolia/react-recommendations/pull/6/commits/956efe16d5ec7ded275cbac4540230c44dd0904c

File: packages/js-recommendations/rollup.config.js (hunk 6 above)
Comment (shortcuts) [reply packages/js-recommendations/rollup.config.js]: Done in https://github.com/algolia/react-recommendations/pull/6/commits/956efe16d5ec7ded275cbac4540230c44dd0904c

File: packages/js-recommendations/rollup.config.js (hunk 5 above)
Comment (shortcuts) [reply packages/js-recommendations/rollup.config.js]: Done in https://github.com/algolia/react-recommendations/pull/6/commits/956efe16d5ec7ded275cbac4540230c44dd0904c

File: package.json (hunk 4 above)
Comment (shortcuts) [reply package.json]: I forgot to update it after my tests but it behave the same way, updated in https://github.com/algolia/react-recommendations/pull/6/commits/956efe16d5ec7ded275cbac4540230c44dd0904c

File: package.json (hunk 3 above)
Comment (shortcuts) [reply package.json]: Done in https://github.com/algolia/react-recommendations/pull/6/commits/47398ad5fd3e21ff8e9edad6b2a0ea0eeb4b9444

File: package.json (hunk 2 above)
Comment (shortcuts) [reply package.json]: Done in https://github.com/algolia/react-recommendations/pull/6/commits/47398ad5fd3e21ff8e9edad6b2a0ea0eeb4b9444

**Review by francoischalifour — APPROVED** [review]

**Pull request merged**